## Στοιχεία Ομάδας

Αριθμός Ομάδας: Μ.Α.11

>  Γκαμουλάκος Παρασκευάς:

>  Καρυπιάδης Ευστράτιος: 03114911

>  Μήτρο Νικόλαος:

---

## Εισαγωγή του Dataset

Το dataset μας προκύπτει έπειτα απο την εφαρμογή του συστήματος _HIPS (Hierarchical Image Processing System)_ σε σειρά εικόνων οχημάτων, οι οποίες ελήφθησαν για τεσσερις διαφορετικές κατηγορίες οχημάτων κάτω από τις ίδιες συνθήκες φωτισμού και γωνίας λήψης. Τα χαρακτηριστικά που εξήχθησαν αντιστοιχούν σε μετρικές ψηφιακής περιγραφής των εικόνων και αποτελούν τις στήλες των δεδομένων μας. Εκμεταλλευόμενοι τα παραπάνω χαρακτηριστικά θα επιχειρήσουμε να ταξινομήσουμε τα δείγματα που έχουμε στην διάθεσή μας στην αντίστοιχη κατηγορία οχήματος που ανήκουν.

Τα αρχεία δεδομένων μας είναι χωρισμένα σε εννέα διαφορετικά αρχεία τα οποία θα ομαδοποιήσουμε σε έναν πίνακα. Αυτό επιτυγχάνεται με το σειριακό διάβασμα των αρχείων και με την τελική ενοποίηση όλων των δεδομένων σε ένα Dataframe. Δεν υπάρχει αρίθμηση για κάθε διαφορετικό δείγμα-γραμμή, αλλά ούτε ονομασία για τα χαρακτηριστικά-στήλες μας. Για αυτό το λόγο επιλέγουμε να ονοματίσουμε της στήλες των δεδομένων μας, συμβουλεύομενοι την περιγραφή του dataset.

In [1]:
import glob
import pandas as pd

path = 'data'
all_data = glob.glob(path + '/*.dat')

l = []

column_names = ['compactness', 'circularity', 'distance_circularity', 'radius_ratio', 'pr.axis_aspect_ratio',
                'max.length_aspect_ratio', 'scatter_ratio', 'elongatedness', 'pr.axis_rectangularity', 
                'max.length_rectangularity', 'major.axis_scaled_variance', 'minor.axis_scaled_variance',
                'gyration_scaled_radius', 'major.axis skewness', 'minor.axis_skewness', 'minor.axis_kurtosis',
                'major.axis_curtosis', 'hollows_ratio', 'label']

for filename in all_data:
    df = pd.read_csv(filename, header=None, sep=' ', names = column_names, usecols= [num for num in range(19)])
    l.append(df)
    
cars = pd.concat(l, axis=0, ignore_index=True)
cars.head(10)

compactness  circularity  distance_circularity  radius_ratio  \
0           95           48                    83           178   
1           91           41                    84           141   
2          104           50                   106           209   
3           93           41                    82           159   
4           85           44                    70           205   
5          107           57                   106           172   
6           97           43                    73           173   
7           90           43                    66           157   
8           86           34                    62           140   
9           93           44                    98           197   

   pr.axis_aspect_ratio  max.length_aspect_ratio  scatter_ratio  \
0                    72                       10            162   
1                    57                        9            149   
2                    66                       10            207   
3                    63                        9            144   
4                   103                       52            149   
5                    50                        6            255   
6                    65                        6            153   
7                    65                        9            137   
8                    61                        7            122   
9                    62                       11            183   

   elongatedness  pr.axis_rectangularity  max.length_rectangularity  \
0             42                      20                        159   
1             45                      19                        143   
2             32                      23                        158   
3             46                      19                        143   
4             45                      19                        144   
5             26                      28                        169   
6             42                      19                        143   
7             48                      18                        146   
8             54                      17                        127   
9             36                      22                        146   

   major.axis_scaled_variance  minor.axis_scaled_variance  \
0                         176                         379   
1                         170                         330   
2                         223                         635   
3                         160                         309   
4                         241                         325   
5                         280                         957   
6                         176                         361   
7                         162                         281   
8                         141                         223   
9                         202                         505   

   gyration_scaled_radius  major.axis skewness  minor.axis_skewness  \
0                     184                   70                    6   
1                     158                   72                    9   
2                     220                   73                   14   
3                     127                   63                    6   
4                     188                  127                    9   
5                     264                   85                    5   
6                     172                   66                   13   
7                     164                   67                    3   
8                     112                   64                    2   
9                     152                   64                    4   

   minor.axis_kurtosis  major.axis_curtosis  hollows_ratio label  
0                   16                  187            197   van  
1                   14                  189            199   van  
2                    9                  188            196  saab  
3                   10 

Στη συνέχεια θα εξάγουμε χρήσιμες πληροφορίες για τα δεδομένα μας:

In [2]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 846 entries, 0 to 845
Data columns (total 19 columns):
compactness                   846 non-null int64
circularity                   846 non-null int64
distance_circularity          846 non-null int64
radius_ratio                  846 non-null int64
pr.axis_aspect_ratio          846 non-null int64
max.length_aspect_ratio       846 non-null int64
scatter_ratio                 846 non-null int64
elongatedness                 846 non-null int64
pr.axis_rectangularity        846 non-null int64
max.length_rectangularity     846 non-null int64
major.axis_scaled_variance    846 non-null int64
minor.axis_scaled_variance    846 non-null int64
gyration_scaled_radius        846 non-null int64
major.axis skewness           846 non-null int64
minor.axis_skewness           846 non-null int64
minor.axis_kurtosis           846 non-null int64
major.axis_curtosis           846 non-null int64
hollows_ratio                 846 non-null int64
label        

In [3]:
cars.describe()

compactness  circularity  distance_circularity  radius_ratio  \
count   846.000000   846.000000            846.000000    846.000000   
mean     93.678487    44.861702             82.088652    168.940898   
std       8.234474     6.169866             15.771533     33.472183   
min      73.000000    33.000000             40.000000    104.000000   
25%      87.000000    40.000000             70.000000    141.000000   
50%      93.000000    44.000000             80.000000    167.000000   
75%     100.000000    49.000000             98.000000    195.000000   
max     119.000000    59.000000            112.000000    333.000000   

       pr.axis_aspect_ratio  max.length_aspect_ratio  scatter_ratio  \
count            846.000000               846.000000     846.000000   
mean              61.693853                 8.567376     168.839243   
std                7.888251                 4.601217      33.244978   
min               47.000000                 2.000000     112.000000   
25%               57.000000                 7.000000     146.250000   
50%               61.000000                 8.000000     157.000000   
75%               65.000000                10.000000     198.000000   
max              138.000000                55.000000     265.000000   

       elongatedness  pr.axis_rectangularity  max.length_rectangularity  \
count     846.000000              846.000000                 846.000000   
mean       40.933806               20.582742                 147.998818   
std         7.811560                2.592138                  14.515652   
min        26.000000               17.000000                 118.000000   
25%        33.000000               19.000000                 137.000000   
50%        43.000000               20.000000                 146.000000   
75%        46.000000               23.000000                 159.000000   
max        61.000000               29.000000                 188.000000   

       major.axis_scaled_variance  minor.axis_scaled_variance  \
count                  846.000000                  846.000000   
mean                   188.625296                  439.911348   
std                     31.394837                  176.692614   
min                    130.000000                  184.000000   
25%                    167.000000                  318.250000   
50%                    178.500000                  364.000000   
75%                    217.000000                  587.000000   
max                    320.000000                 1018.000000   

       gyration_scaled_radius  major.axis skewness  minor.axis_skewness  \
count               846.00000           846.000000           846.000000   
mean                174.70331            72.462175             6.377069   
std                  32.54649             7.486974             4.918353   
min                 109.00000            59.000000             0.000000   
25%                 149.00000            67.000000             2.000000   
50%                 173.00000            71.500000             6.000000   
75%                 198.00000            75.000000             9.000000   
max                 268.00000           135.000000            22.000000   

       minor.axis_kurtosis  major.axis_curtosis  hollows_ratio  
count           846.000000           846.000000     846.000000  
mean             12.599291           188.932624     195.632388  
std               8.931240             6.163949       7.438797  
min               0.000000           176.000000     181.000000  
25%               5.000000           184.000000     190.250000  
50%              11.000000           188.000000     197.000000  
75%              19.000000           193.000000     201.000000  
max              41.000000           206.000000     211.000000

Παρατηρούμε ότι έχουμε στη διάθεσή μας 846 δείγματα, καθένα από τα οποία συνοδεύεται από 18 συνολικά χαρακτηριστικά. Κατά συνέπεια ο πίνακας των δεδομένων μας αποτελείται από 19 στήλες, δεδομένου ότι η τελευταία αντιστοιχεί στις ετικέτες των κατηγοριών. Τα χαρακτηριστικά αυτά είναι αριθμοί, ενώ δεν υπάρχει στήλη στην οποία να απουσιάζουν τιμές. Σημειώνουμε ότι η στήλη των ετικετών περιέχει τιμές σε μορφή κειμένου, τις οποίες όμως δε χρειάζεται να κωδικοποιήσουμε σε αριθμούς, εφόσον είναι διαφορετικές μεταξύ τους.

In [4]:
cars['label'].value_counts(normalize=True) * 100

bus     25.768322
saab    25.650118
opel    25.059102
van     23.522459
Name: label, dtype: float64

Οι κατηγορίες των οχημάτων μας είναι 4 και φέρουν τα ονόματα: _Bus, Van, Opel_ και _Saab_. Έχουμε ισορροπία στο σύνολο των δεδομένων, αφού όπως υπολογίζουμε κάθε κατηγορία συναστά περίπου το 1/4 του Dataset.

---

## Διαχωρισμός Δεδομένων Εκπαίδευσης & Ελέγχου

Αν και οι κατηγορίες μας είναι ισορροπημένες στο σύνολο των δεδομένων μας, για το διαχωρισμό αυτών θα χρησιμοποιήσουμε τη _StratifiedShuffleSplit_ ώστε να εξασφαλίσουμε ότι τα δεδομένα εκπαίδευσης θα αποτελούνται από αντιπροσωπευτικό πλήθος κατηγοριών με το ολικό Dataset:

In [5]:
import numpy as np
from sklearn.model_selection import StratifiedShuffleSplit

split = StratifiedShuffleSplit(n_splits = 1, test_size = 0.2)
for train_index, test_index in split.split(cars, cars['label']):
    train_set = cars.reindex(train_index)
    test_set = cars.reindex(test_index)
    
train_set['label'].value_counts() / len(train_set) * 100

bus     25.739645
saab    25.591716
opel    25.147929
van     23.520710
Name: label, dtype: float64

Τέλος, θα διαχωρίσουμε τα χαρακτηριστικά των δεδομένων μας από την κατηγορία στην οποία ανήκουν:

In [8]:
# Training Set
cars_train = train_set.drop('label', axis=1)
cars_train_labels = train_set['label'].copy()

# Test Set
cars_test = test_set.drop('label', axis=1)
cars_test_labels = test_set['label'].copy()

In [214]:
#cars_train = cars_train_set.values
#cars_train_labels_new = cars_train_labels.values

---

## Baseline Classification

Προτού ξεκινήσουμε την εκπαίδευση των μοντέλων μας θα δημιουργήσουμε τις συναρτήσεις _Cross_Validation_Dummy_ και _Cross_Validation_KNeighbors_ τις οποίες θα χρησιμοποιήσουμε για εκτίμηση της ακρίβειας των αντίστοιχων ταξινομητών που πρόκειται να εκπαιδεύσουμε.

In [9]:
# Source: Hands-On Machine Learning with Scikit-Learn & Tensorflow by Aurélien Géron

from sklearn.base import clone
from sklearn.dummy import DummyClassifier
from sklearn.model_selection import StratifiedKFold

def cross_validation_dummy(data_prepared, data_labels, kfold, w_strat='stratified'):
    
    dc_default = DummyClassifier(strategy=w_strat) # default argument: stratified
    
    result = []
    skfolds = StratifiedKFold(n_splits=kfold, random_state=42)
    
    for train_index, test_index in skfolds.split(data_prepared, data_labels):
        clone_clf = clone(dc_default)
        data_train_folds = data_prepared[train_index]
        data_train_labels_folds = (data_labels[train_index])
        data_test_fold = data_prepared[test_index]
        data_test_lables_fold = (data_labels[test_index])

        clone_clf.fit(data_train_folds, data_train_labels_folds)
        y_pred = clone_clf.predict(data_test_fold)
        n_correct = sum(y_pred == data_test_labels_fold)
        result.append(n_correct / len(y_pred))
    
    return result

- #### DummyClassifier

Αρχικά θα εκπαιδεύσουμε έναν _DummyClassifier_ χωρίς να μεταβάλλουμε τις παραμέτρους αρχικοποίησής του:

In [ ]:

from sklearn.metrics import accuracy_score

dc_default = DummyClassifier() # default strategy: stratified
dc_default.fit(cars_train_set, cars_train_labels) # fitting on train_set

cars_predictions = dc_default.predict(cars_test_set) # predicting on test_set

target_names = dc_default.classes_.tolist() # fetch class names

accuracy_score(cars_test_labels, cars_predictions) * 100

Παρατηρούμε ότι η ακρίβεια του _DummyClassifier_ είναι αρκετά χαμηλή, με ποσοστό περίπου 21%.

Συνεχίζοντας, θα πάρουμε περισσότερες πληροφόριες για τη διαδικασία ταξινόμησης, υπολογίζοντας _Confusion Table_ καθώς και τις τιμές _Precision, Recall_ και _F1-Score_.

In [ ]:
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt
from matplotlib import cm

matrix = confusion_matrix(cars_test_labels, cars_predictions)

fig = plt.figure(figsize=(8,6))
ax = fig.add_subplot(111)
cax = ax.matshow(matrix, cmap=cm.get_cmap('PuBu'))
fig.colorbar(cax)

ax.set_xticklabels([''] + target_names)
ax.set_yticklabels([''] + target_names)

plt.show()

In [ ]:
plt.clf()

In [ ]:
from sklearn.metrics import f1_score
from sklearn.metrics import classification_report
from sklearn.metrics import precision_recall_fscore_support

dc_f1_micro = f1_score(cars_test_labels, cars_predictions, average='micro')
dc_f1_macro = f1_score(cars_test_labels, cars_predictions, average='macro')

print("F1-Micro Average: " + str(dc_f1_micro))
print("F1-Macro Average: " + str(dc_f1_macro) + "\n")

print(classification_report(cars_test_labels, cars_predictions))

Μελετώντας τον πίνακα σύγχυσης διασταυρώνουμε το ποσοστό επιτυχίας που σημειώσαμε παραπάνω, παρατηρώντας πως η κύρια διαγώνιος 
δεν παρουσιάζει εμφανώς έντονες τιμές. Σχετικά με τα _F1-Macro_ και _F1-Micro_ αναμένουμε να έχουν σχεδόν ίδιες τιμές, επειδή το πλήθος των δεδομένων μας ήταν ισορροπημένο ως προς τις κατηγορίες οχημάτων και κατά συνέπεια δε δημιουργείται κάποια κλίση ως προς κάποια κατηγορία. Αξίζει να σημειώσουμε ότι οι κατηγορίες _Bus_ και _Van_ εμφανίζουν τα υψηλότερα ποσοστά ακρίβειας και ανάκλισης. Για παράδειγμα, εάν το σύστημά μας προβλέψει πως μία εικόνα είναι λεωφορείο είναι σωστό με ποσοστο 23%, ενώ για τις συνολικές λήψεις φορτηγών είναι ικανό να ανιχνεύει μόλις το 25% αυτών. Η παραπάνω παρατήρηση είναι αναμενόμενη, εάν λάβουμε υπόψην ότι οι δύο αυτές κατηγορίες οχημάτων είναι αισθητά διαφορετικές τόσο μεταξύ τους, όσο και με τα παραδοσιακά αμάξια _0pel_ και _Saab_ καθιστώντας το διαχωρισμό τους περισσότερο προφανή.

- #### KNeighborsClassifier

Θα συνεχίσουμε με την εκπαίδευση ενός προεπιλεγμένου _KNeighborsClassifier_ :

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier() # default n: 5
knn.fit(cars_train_set, cars_train_labels) # fitting on train_set

cars_predictions = knn.predict(cars_test_set) # predicting on test_set

target_names = knn.classes_.tolist() # fetch class names

accuracy_score(cars_test_labels, cars_predictions) * 100

Η ακρίβεια σε αυτήν την περίπτωση είναι αισθητά υψηλότερη από προηγουμένως. Ας εξετάσουμε τις ανάλογες μετρικές απόδοσης:

In [ ]:
matrix = confusion_matrix(cars_test_labels, cars_predictions)

fig = plt.figure(figsize=(8,6))
ax = fig.add_subplot(111)
cax = ax.matshow(matrix, cmap=cm.get_cmap('PuBu'))
fig.colorbar(cax)

ax.set_xticklabels([''] + target_names)
ax.set_yticklabels([''] + target_names)

plt.show()

In [ ]:
plt.clf()

In [ ]:
kn_f1_micro = f1_score(cars_test_labels, cars_predictions, average='micro')
kn_f1_macro = f1_score(cars_test_labels, cars_predictions, average='macro')

print("F1-Micro Average: " + str(kn_f1_micro))
print("F1-Macro Average: " + str(kn_f1_macro) + "\n")

print(classification_report(cars_test_labels, cars_predictions))

Ο πίνακας σύγχυσης για το δεδομένο ταξινομητή παρατηρούμε ότι είναι περισσότερο διαγωνιοποιημένος, με υψηλά ποσοστά ακρίβειας για τις κατηγορίες _Bus_ και _Van_ , όπως αναμέναμε. Ανάλογη αύξηση σημειώνουμε και στις τιμές του F1-score, με παραπλήσιες τιμές για Macro και Micro.
Τέλος σημειώνουμε ότι αν και ο ταξινομητής παρουσιάζει βελτιωμένη απόδοση με προηγουμένως, αντιμετωπίζει και αυτός δυσκολία στο διαχωρισμό των δύο παραδοσιακών αυτοκινήτων _Opel_ και _Saab_ , δεδομένου ότι μπορεί να αναγνωρίσει μονάχα το 38% και 45% αυτών, αντίστοιχα.

Ολοκληρώντας την πρώτη μελέτη των δεδομένων μας με τους _Dummy_ και _KNeighbors_ ταξινομητές μας, παρουσιάζουμε παρακάτω ένα διάγραμμα συγκρίνοντας τις τιμές των βαθμολογιών F1 ξεχωριστά για κάθε ταξινομητή:

In [ ]:
labels = ['DummyClassifier', 'KNeighbors']
f1_micro = [round(dc_f1_micro,3), round(kn_f1_micro,3)]
f1_macro = [round(dc_f1_macro,3), round(kn_f1_macro,3)]

x = np.arange(len(labels))  # the label locations
width = 0.25  # the width of the bars

fig, ax = plt.subplots(figsize=(8,6))
rects1 = ax.bar(x - width/2, f1_micro, width, label='F1-Micro Score', color=(0.2, 0.4, 0.6, 0.3))
rects2 = ax.bar(x + width/2, f1_macro, width, label='F1-Macro Score', color=(0.2, 0.4, 0.6, 0.9))

ax.set_ylabel('Scores')
ax.set_title('F1-Score Comparison')
ax.set_xticks(x)
ax.set_xticklabels(labels)
ax.legend()

def autolabel(rects):
    """Attach a text label above each bar in *rects*, displaying its height."""
    for rect in rects:
        height = rect.get_height()
        ax.annotate('{}'.format(height),
                    xy=(rect.get_x() + rect.get_width() / 2, height),
                    xytext=(0, 3),  # 3 points vertical offset
                    textcoords="offset points",
                    ha='center', va='bottom')

autolabel(rects1)
autolabel(rects2)

fig.tight_layout()

plt.show()

---

## Βελτιστοποίηση Ταξινομητών

Προς βελτίωση των ταξινομητών μας θα ασχοληθούμε αρχικά με την προεπεξεργασία των δεδομένων:

- ### StandardScaler

In [ ]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()

cars_train_set_scaled = scaler.fit_transform(cars_train_set)
cars_test_set_scaled = scaler.transform(cars_test_set)

- ### PCA

In [ ]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.99)

cars_train_set_scaled_reduced = pca.fit_transform(cars_train_set_scaled)
cars_test_set_scaled_reduced = pca.transform(cars_test_set_scaled)

In [ ]:
cars_train_set_scaled_reduced.shape

In [ ]:
knn = KNeighborsClassifier(n_neighbors=5)

knn.fit(cars_train_set_scaled_reduced, cars_train_labels) # fitting on train_set

cars_predictions = knn.predict(cars_test_set_scaled_reduced) # predicting on test_set

accuracy_score(cars_test_labels, cars_predictions) * 100

In [ ]:
dc_default.fit(cars_train_set_scaled_reduced, cars_train_labels) # fitting on train_set

cars_predictions = dc_default.predict(cars_test_set_scaled_reduced) # predicting on test_set

accuracy_score(cars_test_labels, cars_predictions) * 100